src/mlpack/methods/ann/layer/maxout_impl.hpp

```diff
@@ -0,0 +1,134 @@
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "maxout.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout(
+    const size_t inSize,
+    const size_t hiddenSize,
+    const size_t outSize) :
+    inSize(inSize),
+    hiddenSize(hiddenSize),
+    outSize(outSize)
+{
+  weights.set_size((inSize * hiddenSize * outSize) + (outSize * hiddenSize), 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void Maxout<InputDataType, OutputDataType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), outSize * hiddenSize, inSize,
+      false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize * hiddenSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.n_elem == 0)
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+
+  hiddenParameter = (weight * input) + bias;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+    output.row(j) = arma::max(hiddenParameter.rows(i, i + hiddenSize - 1), 0);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
+          hiddenParameterSubset.col(k)) == hiddenParameterSubset.col(k)));
+      hiddenGradient(i + ind, k) = gy(j, k);
+    }
+  }
+
+  g = weight.t() * hiddenGradient;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Gradient(
+    const arma::Mat<eT>&& input,
+    arma::Mat<eT>&& error,
+    arma::Mat<eT>&& gradient)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
```

Comment:
I used following code in **maxoutIndices**

in forward pass I used following loop.
```
for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++) {
    hiddenParameterSubset = hiddenParameter.submat(i, 0, i + hiddenSize - 1, 0);
    output.row(j) = arma::max(hiddenParameterSubset, 0);

    for (int k = 0; k < output.n_cols; k++) {
      maxoutIndices(j, k) = arma::as_scalar(
          arma::find(hiddenParameterSubset == output(j, k)));
    }
```

Calculating index in backward and gradient method.
```
int ind = maxoutIndices(j, k);
```

Similarly in **maxoutValue**

forward:
```
for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++) {
    hiddenParameterSubset = hiddenParameter.submat(i, 0, i + hiddenSize - 1, 0);
    output.row(j) = arma::max(hiddenParameterSubset, 0);
    maxoutValues.row(j) = arma::max(hiddenParameterSubset, 0);
}
```

backward and gradient:
```
int ind = arma::as_scalar(arma::find(maxoutValues(j, k) == hiddenParameterSubset.col(k)));
```

Well initially I was doing timing calculations to measure how much boost I am able to get from these methods but then I did not get any improvement so I ran tests again and again for 3 times and I still got same results. Therefore I posted them here because I am confused from these results.

Also I used thyroid dataset for bench marking. I don't think that these test results are influenced by dataset, are they ?